### PR TITLE
Test connector templates

### LINKED ISSUE #725: Test that template generation produces valid modules
## Tell us about the problem you're trying to solve
#700 introduced template generation for java destinations. However, since that change now houses templates instead of actual code, we can't verify that the output of that template is valid java code. The task here is to add a test suite which verifies that the output of that module is valid code.

### PR DESCRIPTION
**Pull request by michel-tricot**
## What
closes #725 

## How
Scaffold python source template
Fix current scaffold that wasn't working
Add an extra option for the singer scaffold

## Recommended reading order
1. `airbyte-integrations/connector-templates/generator/build.gradle`
1. `airbyte-integrations/connector-templates/generator/plopfile.js`
1. the rest

### CODE REVIEW

**Review by cgardens — APPROVED** [review]
I want to repeat back my understanding of how this works:

* as part of the build process, gradle runs add scaffold task `source-python`, which overwrites any existing output of the scaffold.
* it then makes sure that the output of the template builds properly.
* if there are changes to the template that haven't been committed, then the build will fail because there will be a diff

That all looks good to me!

File: airbyte-integrations/connector-templates/source-singer/source_{{snakeCase name}}_singer/source.py.hbs [hunk 1]
```diff
@@ -30,7 +30,7 @@ from base_singer import SingerSource, Status
 
 
 class Source{{properCase name}}Singer(SingerSource):
-    TAP_CMD = "tap-{{properCase name}}"
+    TAP_CMD = "{{ tap_name }}"
```
Comment (sherifnada) [inline airbyte-integrations/connector-templates/source-singer/source_{{snakeCase name}}_singer/source.py.hbs:33]: this is not always the case e.g: sometimes pypi package `pipelinewise-tap-google-analytics` is called via `tap-google-analytics`

File: airbyte-integrations/connector-templates/source-singer/setup.py.hbs [hunk 2]
```diff
@@ -38,7 +38,7 @@ setup(
         # Dependencies required by the main package but not tests should go in main. Deps required by
         # integration tests but not the main package go in tests. Deps required by both should go in
         # install_requires.
-        "main": ["base-singer", "base-python"],
+        "main": ["{{ tap_name }}", "base-singer", "base-python"],
```
Comment (sherifnada) [inline airbyte-integrations/connector-templates/source-singer/setup.py.hbs:41]: can we get rid of the `main` package and put all this in `install_requires`?

Comment (sherifnada) [reply airbyte-integrations/connector-templates/source-singer/setup.py.hbs:41]: also, we should get the tap version from them so it doesn't change from underneath us

**Review by sherifnada — APPROVED** [review]
Nice!

**Comment by sherifnada** [comment]
@michel-tricot maybe try making this scaffold step depend on generating the airbyte protocol module as well? if that doesn't work could you share a build scan of the failing build?

**Comment by michel-tricot** [comment]
@cgardens that is correct

File: airbyte-integrations/connector-templates/source-singer/setup.py.hbs (hunk 2 above)
Comment (michel-tricot) [reply airbyte-integrations/connector-templates/source-singer/setup.py.hbs:41]: to be clear my goal is not to make that part easier for a user. Just to be able to inject an existing tap to ensure our implementation continues to work

File: airbyte-integrations/connector-templates/source-singer/source_{{snakeCase name}}_singer/source.py.hbs (hunk 1 above)
Comment (michel-tricot) [reply airbyte-integrations/connector-templates/source-singer/source_{{snakeCase name}}_singer/source.py.hbs:33]: same. My goal isn't to fix something that wasn't working and that can't really be "templated". I just want to enable automatic testing

**Comment by michel-tricot** [comment]
I changed the testing strategy. 

I had enabled the regeneration at gradle configuration time because gradle didn't like to have files removed during execution. This was nasty: clean would generate the scaffold and build was unstable.

Instead I rely the fact that the generated template is a valid integration and should have the same lifecyle. On every PR build I regen the code. if the code changes then the build will fail: it should only happen if you have played with the templates, in which case you should regen it (with the gradle task).

I have disabled singer test as i would need a "noop" singer tap to test and I don't have one at the moment

**Pull request merged**